### PR TITLE
Mem intg testing

### PR DESCRIPTION
**Pull request by GregAC**
This is a mix @ctopal's work in https://github.com/lowRISC/ibex/pull/1811 and work I've done. It supersedes https://github.com/lowRISC/ibex/pull/1811

### CODE REVIEW

**Review by ctopal — APPROVED** [review]
LGTM

File: dv/cosim/spike_cosim.cc [hunk 1]
```diff
@@ -376,6 +377,12 @@ bool SpikeCosim::check_sync_trap(uint32_t write_reg,
     return false;
   }
 
+  // If we see an internal NMI, that means we receive an extra memory intf item.
+  // Deleting that is necessary since next Load/Store would fail otherwise.
+  if (processor->get_state()->mcause->read() == 0xFFFFFFE0) {
```
Comment (marnovandermaas) [inline dv/cosim/spike_cosim.cc:409]: Why don't we check the `nmi_int` value instead of `mcause`?

File: dv/cosim/spike_cosim.cc [hunk 2]
```diff
@@ -276,7 +295,13 @@ bool SpikeCosim::step(uint32_t write_reg, uint32_t write_reg_data,
   }
   pending_iside_error = false;
 
-  if (!check_retired_instr(write_reg, write_reg_data, pc)) {
+  if (suppress_reg_write) {
+    // If we suppressed a register write restore the old register state now
+    processor->get_state()->XPR.write(suppressed_write_reg,
+                                      suppressed_write_reg_data);
+  }
```
Comment (marnovandermaas) [inline dv/cosim/spike_cosim.cc:302]: Isn't there a way to stop the write in the first place instead of undoing it?

File: dv/cosim/spike_cosim.cc [hunk 3]
```diff
@@ -913,4 +965,44 @@ bool SpikeCosim::check_debug_ebreak(uint32_t write_reg, uint32_t pc,
...
+      // Not valid load encodings
+      return false;
+    }
+
+    rd_out = (insn_32 >> 7) & 0x1F;
+    return true;
+  }
+
+  return false;
+}
```
Comment (marnovandermaas) [inline dv/cosim/spike_cosim.cc:1006]: I don't like that this replicates decode logic, although I don't have a concrete alternative in mind.

**Pull request merged**